diff --git a/serie3/compile.c b/serie3/compile.c
index 6143967..ff74a18 100644
--- a/serie3/compile.c
+++ b/serie3/compile.c
@@ -1,8 +1,7 @@
-/* TODO: Task (b) Please fill in the following lines, then remove this line.
- *
- * author(s):   FIRSTNAME LASTNAME 
- *              (FIRSTNAME2 LASTNAME2)
- * modified:    2010-01-07
+/* 
+ * author(s):   Mathieu Simon
+ *              Thomas Rickenbach
+ * modified:    2011-04-20
  *
  */
 
@@ -13,7 +12,20 @@
 #include "compiler.h"
  
 int main ( int argc, char** argv ) {
-    /* TODO: Task (c) implement main */
-    return EXIT_SUCCESS;
+	int count;
+	count = argc;
+	count = count-1;
+
+	if (count == 2){
+		printf("Input:    %s\n",argv[1]);
+		printf("Postfix:  ");
+                compiler(argv[1], argv[2]);
+		return EXIT_SUCCESS;
+	}
+
+	else {
+		printf("usage: <commandName> expression filename\n");
+		return EXIT_FAILURE;
+        }
 }
 
