From a6d4ecc6272947a83eb03c315cafa70353caed5d Mon Sep 17 00:00:00 2001
From: Mathieu Simon <mathieu.simon@simweb.ch>
Date: Thu, 21 Apr 2011 12:03:25 +0200
Subject: [PATCH] memory.c: Some part maybe works

---
 serie3/memory.c |   45 +++++++++++++++++++++++++++++++++++++++------
 1 files changed, 39 insertions(+), 6 deletions(-)

diff --git a/serie3/memory.c b/serie3/memory.c
index 875bac0..1ac2491 100644
--- a/serie3/memory.c
+++ b/serie3/memory.c
@@ -1,8 +1,7 @@
-/* TODO: Task (b) Please fill in the following lines, then remove this line.
- *
- * author(s):   FIRSTNAME LASTNAME 
- *              (FIRSTNAME2 LASTNAME2)
- * modified:    2010-01-07
+/*
+ * author(s):   Thomas Rickenbach
+ *              Mathieu Simon
+ * modified:    2011-04-20
  *
  */
 
@@ -129,7 +128,41 @@ void initializeMemory() {
 
 /* Load a file to memory */
 void loadFile(char* filename) {
-    /* TODO: Task (d) implement loadFile */
+	char *buffer;
+	unsigned long fileLen;
+	int i;
+   	word wo;
+	
+	/* Open file */
+	outputStream = fopen(filename, "rb");
+	if (!outputStream){
+		fprintf(stderr, "Unable to open file %s", filename);
+		return;
+	}
+
+	/* Detect file length */
+	fseek(outputStream, 0, SEEK_END);
+	fileLen=ftell(outputStream);
+	fseek(outputStream, 0, SEEK_SET);
+
+	buffer=(char *)malloc(fileLen+1);
+	if (!buffer){
+		fprintf(stderr, "Memory error!");
+                fclose(outputStream);
+		return;
+	}
+
+	/* Read file content in buffer */
+	fread(buffer, fileLen, 1, outputStream);
+	fclose(outputStream);
+
+	/* Cut buffer in single words to memory */
+	for (i=0; i<sizeof(buffer)-1; i++) {
+	        wo = (buffer[i] >> (8*(4-(i+1)))) & 0xFF;
+		storeWord(wo, memoryRegionCounter+i);
+    	}
+
+	free(buffer);
 }
 
 /* ========================================================================== */
-- 
1.7.1

